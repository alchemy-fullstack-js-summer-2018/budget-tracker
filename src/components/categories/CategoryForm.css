:local(.categoryForm) {

  display: grid;
  height: 80px;
  grid-template: 20px 1fr / 1fr 124px;
  grid-template-areas:
    ". x"
    "inputs buttons";
  border: 1px solid lightgrey;
  padding: 0px 20px 20px 20px;
  
  label {
    display: block;
  }

  

  input {
    border: 1px solid rgb(226, 145, 23);
    height: 25px;
    margin: 1px;
  }
  
  .inputs {
    grid-area: inputs;
  }

  .buttons {
    grid-area: buttons;
    display: flex;
    justify-content: space-around;
    align-items: center;
  }

  button {
    border: 1px solid rgb(226, 145, 23);
    font-size: 20px;
    width: 60px;
    height: 60px;
    background: white;
  }

  .fa-ban {
    color: red;
  }

  .fa-plus, .fa-check {
    color: green
  }

  .fa-times {
    color: red;
  }

  .fa-times {
    grid-area: x;
    display: flex;
    justify-content: flex-end;
    align-items: start;
  }
}