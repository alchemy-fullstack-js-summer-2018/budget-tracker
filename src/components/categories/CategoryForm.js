import React, { Component } from 'react';
import PropTypes from 'prop-types';

class CategoryForm extends Component {

  state = {
    editing: false,
    key: null,
    timestamp: new Date(),
    name: '',
    budget: 0
  };

  static propTypes = {
    category: PropTypes.object,
    onComplete: PropTypes.func.isRequired,
    onCancel: PropTypes.func
  };

  componentDidMount() {
    const { category } = this.props;
    if(!category) return;

    this.setState(category);
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { key, name, budget, timestamp } = this.state;
    const category = { name, budget, timestamp }
;
    if(key) category.key = key;

    const { onComplete, category: originalCategory } = this.props;

    onComplete(category)
      .then(() => {
        if(!originalCategory) {
          this.setState({ name: '', budget: null });
          document.activeElement.blur();
        }
      })
      .catch(err => {
        this.setState({ error: err });
      });
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() {
    const { key, name, budget } = this.state;
    const { onCancel } = this.props;

    return (
      <form onSubmit={this.handleSubmit}>
        <InputControl name="name" value={name} onChange={this.handleChange}/>
        <InputControl name="budget" value={budget} onChange={this.handleChange}/>
        <p>
          <button type="submit">{ key ? 'Update' : 'Add' }</button>
          {key && <button type="button" onClick={onCancel}>Cancel</button>}
        </p>
      </form>
    );
  }
}

const InputControl = (props) => (
  <p>
    <label>
      {props.name}:
      <input {...props} required/>
    </label>
  </p>
);

export default CategoryForm;