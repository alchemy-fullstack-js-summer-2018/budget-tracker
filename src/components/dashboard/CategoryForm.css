:local(.categoryForm) {
  display: flex;
  flex-direction: column;
  align-items: center;
}