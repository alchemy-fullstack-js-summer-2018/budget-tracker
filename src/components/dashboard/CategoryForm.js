import React, { Component } from 'react';
import PropTypes from 'prop-types';

class CategoryForm extends Component {

  state = {
    id: null,
    timestamp: '',
    name: '',
    budget: ''
  };

  static propTypes = {
    category: PropTypes.object,
    onComplete: PropTypes.func.isRequired
  };

  componentDidMount() {
    const { category } = this.props;
    if(!category) return;

    this.setState(category);
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { name, budget, timestamp, id } = this.state;
    const category = { name, budget, timestamp };
    if(id) category.id = id;

    this.props.onComplete(category);
    this.setState({ name: '', budget: '', timestamp: '' });
  };

  handleChange = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  render() {
    const { id, name, budget } = this.state;

    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Category:&nbsp;<br/>
          <input name="name" value={name} onChange={this.handleChange}/>
        </label>
        <br/>
        <label>
          Budget:&nbsp;<br/>
          <input name="budget" value={budget} onChange={this.handleChange}/>
        </label>
        <br/>
        <button type="submit">{id ? 'Update' : 'Add' }</button>
      </form>
    );
  }
}

export default CategoryForm;