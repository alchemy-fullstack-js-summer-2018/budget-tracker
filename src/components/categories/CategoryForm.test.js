import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import CategoryForm from './CategoryForm';

describe('Category form', () => {
  it('renders add if no category prop', () => {
    const handleComplete = jest.fn();
    const wrapper = mount(<CategoryForm onComplete={handleComplete}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();

    const category = { name: 'clothes', key: null, budget: 500 };

    wrapper.find('input[name="name"]').simulate('change', {
      target: {
        name: 'name',
        value: category.name
      }
    });

    wrapper.find('input[name="budget"]').simulate('change', {
      target: {
        name: 'budget',
        value: category.budget
      }
    });

    wrapper.find('button').simulate('submit');

    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toEqual(category);
  });

  it('renders edit if there is category prop', () => {
    const handleComplete = jest.fn();
    const handleCancel = jest.fn();

    const category = { name: 'clothes', key: null, budget: 500 };

    const wrapper = mount(<CategoryForm
      onComplete={handleComplete}
      onCancel={handleCancel}
      category={category}
    />);

    expect(toJSON(wrapper)).toMatchSnapshot();

    wrapper.find('input[name="budget"]').simulate('change', {
      target: {
        name: 'budget',
        value: 200
      }
    });

    wrapper.find('button[type="submit"]').simulate('submit');

    const calls = handleComplete.mock.calls;
    expect(calls.length).toBe(1);
    expect(calls[0][0]).toEqual({
      ...category,
      budget: 200
    });
  });
});