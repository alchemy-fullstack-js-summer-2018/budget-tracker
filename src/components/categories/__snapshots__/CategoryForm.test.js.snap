// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category form renders add if no category prop 1`] = `
<CategoryForm
  onComplete={[MockFunction]}
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="budget"
      onChange={[Function]}
      value={null}
    >
      <p>
        <label>
          budget
          :
          <input
            name="budget"
            onChange={[Function]}
            required={true}
            value={null}
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</CategoryForm>
`;

exports[`Category form renders add if no category prop 2`] = `
<CategoryForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "budget": "400",
            "name": "shopping",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="budget"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          budget
          :
          <input
            name="budget"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</CategoryForm>
`;

exports[`Category form renders add if no category prop 3`] = `
<CategoryForm
  onComplete={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "budget": "400",
            "name": "shopping",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Promise {},
        },
      ],
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <InputControl
      name="name"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          name
          :
          <input
            name="name"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <InputControl
      name="budget"
      onChange={[Function]}
      value=""
    >
      <p>
        <label>
          budget
          :
          <input
            name="budget"
            onChange={[Function]}
            required={true}
            value=""
          />
        </label>
      </p>
    </InputControl>
    <p>
      <button
        type="submit"
      >
        Add
      </button>
    </p>
  </form>
</CategoryForm>
`;
